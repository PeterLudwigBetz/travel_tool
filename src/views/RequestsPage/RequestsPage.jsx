import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import WithLoadingTable from '../../components/Table';
import RequestPanelHeader from '../../components/RequestPanelHeader/RequestPanelHeader';
import Utils from '../../helper/Utils';
import Modal from '../../components/modal/Modal';
import Base from '../Base';
import { NewRequestForm } from '../../components/Forms';
import {
  fetchUserRequests,
  createNewRequest
} from '../../redux/actionCreator/requestActions';
import { openModal, closeModal } from '../../redux/actionCreator/modalActions';
import { fetchRoleUsers } from '../../redux/actionCreator/roleActions';

export class RequestsPage extends Base {
  state = {
    hideNotificationPane: true,
    hideSideBar: false,
    openSearch: false,
    hideNewRequestModal: true,
    selectedLink: 'request page',
    activeStatus: Utils.getActiveStatus(this.props.location.search),
    url: this.props.location.search,
    hideOverlay: false
  };

  componentDidMount() {
    const { fetchUserRequests, fetchRoleUsers } = this.props;
    const { url } = this.state;
    fetchUserRequests(url);
    // Fetch managers
    fetchRoleUsers(53019);
  }

  fetchRequests = query => {
    const { history } = this.props;
    history.push(`/requests${query}`);
  };

  onPageChange = page => {
    const { url } = this.state;
    const query = Utils.buildQuery(url, 'page', page);
    this.fetchRequests(query);
  };

  getRequestsWithLimit = limit => {
    const { url } = this.state;
    let query;
    const pages = Math.ceil(this.props.pagination.dataCount / limit);
    if (pages < this.props.pagination.currentPage) {
      const newUrl = Utils.buildQuery(url, 'page', pages);
      query = Utils.buildQuery(newUrl, 'limit', limit);
    } else {
      query = Utils.buildQuery(url, 'limit', limit);
    }
    this.fetchRequests(query);
  };

  renderRequestPanelHeader() {
    const {
      openRequestsCount,
      requests,
      pastRequestsCount,
      openModal,
      shouldOpen,
      modalType
    } = this.props;
    const { url, activeStatus } = this.state;

    return (
      <div className="rp-requests__header">
        <RequestPanelHeader
          openRequestsCount={openRequestsCount}
          fetchRequests={this.fetchRequests}
          requestsLength={requests.length}
          pastRequestsCount={pastRequestsCount}
          getRequestsWithLimit={this.getRequestsWithLimit}
          url={url}
          activeStatus={activeStatus}
          openModal={openModal}
          shouldOpen={shouldOpen}
          modalType={modalType}
        />
      </div>
    );
  }

  renderRequests(requests, isLoading, error, message) {
    const { openModal, closeModal, shouldOpen, modalType } = this.props;
    return (
      <div className="rp-table">
        <WithLoadingTable
          requests={requests}
          isLoading={isLoading}
          fetchRequestsError={error}
          closeModal={closeModal}
          openModal={openModal}
          shouldOpen={shouldOpen}
          modalType={modalType}
          message={message}
        />
      </div>
    );
  }

  renderNewRequestForm(shouldOpen, modalType) {
    const { user, createNewRequest, loading, errors, closeModal, manager } = this.props;
    return (
      <Modal
        closeModal={closeModal}
        visibility={
          shouldOpen && modalType === 'new model' ? 'visible' : 'invisible'
        }
        title="New Travel Request"
      >
        <NewRequestForm
          user={user}
          handleCreateRequest={createNewRequest}
          loading={loading}
          errors={errors}
          closeModal={closeModal}
          managers={manager}
        />
      </Modal>
    );
  }

  renderRequestPage(
    leftPaddingClass,
    hideClass,
    hideClass3,
    selectedLink,
    hideSideBar
  ) {
    const {
      isLoading,
      requests,
      pagination,
      fetchRequestsError,
      message
    } = this.props;
    return (
      <div className="mdl-layout__content full-height">
        <div className="mdl-grid mdl-grid--no-spacing full-height">
          {this.renderLeftSideBar(hideSideBar, selectedLink)}
          <div className="mdl-cell mdl-cell--9-col-desktop request-page__table-view mdl-cell--8-col-tablet mdl-cell--4-col-phone">
            <div className={`rp-requests ${leftPaddingClass}`}>
              {this.renderRequestPanelHeader()}
              {requests &&
                this.renderRequests(
                  requests,
                  isLoading,
                  fetchRequestsError,
                  message
                )}
              {!isLoading &&
                requests.length > 0 &&
                this.renderPagination(pagination)}
            </div>
          </div>
          {this.renderNotificationPane(hideClass, hideSideBar)}
        </div>
      </div>
    );
  }

  render() {
    const {
      hideNotificationPane,
      hideSideBar,
      selectedLink,
      openSearch,
      hideOverlay,
    } = this.state;
    const { shouldOpen, modalType } = this.props;
    let [hideClass, leftPaddingClass] = hideNotificationPane
      ? ['hide', '']
      : ['', 'pd-left'];
    const hideClass3 = hideSideBar ? '' : 'hide mdl-cell--hide-desktop';
    const overlayClass = hideOverlay ? 'block' : 'none';
    return (
      <div>
        {this.renderOverlay(overlayClass)}
        <div className="mdl-layout mdl-js-layout request-page mdl-layout--no-desktop-drawer-button">
          {this.renderSideDrawer(selectedLink, overlayClass)}
          {this.renderNavBar(openSearch)}
          {this.renderNewRequestForm(shouldOpen, modalType)}
          {this.renderRequestPage(
            leftPaddingClass,
            hideClass,
            hideClass3,
            selectedLink,
            hideSideBar
          )}
        </div>
      </div>
    );
  }
}

RequestsPage.propTypes = {
  user: PropTypes.object,
  fetchUserRequests: PropTypes.func.isRequired,
  fetchRoleUsers: PropTypes.func.isRequired,
  fetchRequestsError: PropTypes.string,
  requests: PropTypes.array,
  pagination: PropTypes.object,
  openRequestsCount: PropTypes.number,
  pastRequestsCount: PropTypes.number,
  isLoading: PropTypes.bool,
  url: PropTypes.string,
  history: PropTypes.shape({}).isRequired,
  createNewRequest: PropTypes.func.isRequired,
  creatingRequest: PropTypes.bool,
  errors: PropTypes.array,
  shouldOpen: PropTypes.bool.isRequired,
  modalType: PropTypes.string,
  openModal: PropTypes.func.isRequired
};

RequestsPage.defaultProps = {
  url: '',
  fetchRequestsError: null,
  requests: [],
  pagination: {},
  isLoading: false,
  creatingRequest: false,
  openRequestsCount: null,
  pastRequestsCount: null,
  errors: [],
  modalType: '',
  user: {}
};

export const mapStateToProps = ({ requests, modal, role }) => ({
  ...requests,
  ...modal.modal,
  ...role
});

const actionCreators = {
  fetchUserRequests,
  createNewRequest,
  fetchRoleUsers,
  openModal,
  closeModal
};

export default connect(
  mapStateToProps,
  actionCreators
)(RequestsPage);
