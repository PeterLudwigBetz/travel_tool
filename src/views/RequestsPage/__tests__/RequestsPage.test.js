import React from 'react';
import sinon from 'sinon';
import RequestsPage from '../RequestsPage';

const props = {
  requests: [
    {
      'id':'745923RTF',
      'destination':'Lagos',
      'origin':'Nairobi',
      'duration':'3 days',
      'startDate':'12 Oct 2018',
      'status':'Open'
    },
    {
      'id':'745923RTF',
      'destination':'New York',
      'origin':'Nairobi',
      'duration':'3 days',
      'startDate':'12 Oct 2018',
      'status':'Rejected'
    },
    {
      'id':'745923RTF',
      'destination':'Kampala',
      'origin':'Nairobi',
      'duration':'3 days',
      'startDate':'12 Oct 2018',
      'status':'Approved'
    }
  ],
  pagination: {
    currentPage: 2,
    pageCount: 4,
    onPageChange: sinon.spy(),
  }
};

const wrapper = shallow(<RequestsPage {...props} />);
const spy = sinon.spy(wrapper.instance(), 'onPageChange');

describe('<RequestsPage>', () => {
  it('should render the Requests and Pagination components successfully', () => {
    expect(wrapper.find('Requests').length).toBe(1);
    expect(wrapper.find('Pagination').length).toBe(1);
  });

  it('calls the onPageChange method', () => {
    wrapper.instance().onPageChange(2);
    expect(spy.calledOnce).toEqual(true);
    expect(spy.calledWith(2)).toEqual(true);
  });
});
