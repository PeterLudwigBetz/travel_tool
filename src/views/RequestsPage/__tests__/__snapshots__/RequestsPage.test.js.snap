// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestsPage> renders as expected 1`] = `
<div>
  <div
    className="side_overlay"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex="0"
  />
  <div
    className="mdl-layout mdl-js-layout request-page mdl-layout--no-desktop-drawer-button"
  >
    <Connect(SideDrawer)
      selectedLink="request page"
      showDrawer="none"
    />
    <withRouter(Connect(NavBar))
      avatar="upic.svg"
      className=""
      handleHideSearchBar={[Function]}
      handleShowDrawer={[Function]}
      onNotificationToggle={[Function]}
      openSearch={false}
    />
    <Modal
      closeModal={[Function]}
      description=""
      divClass=""
      dynamicDate=""
      dynamicText=""
      innerClass=""
      modalBar={<div />}
      nextClass=""
      symbol=""
      title="New Travel Request"
      visibility="invisible"
    >
      <NewRequestForm
        closeModal={[Function]}
        creatingRequest={false}
        errors={Array []}
        handleCreateRequest={[Function]}
        loading={false}
        managers={Array []}
        user={
          Object {
            "UserInfo": Object {
              "name": "John Doe",
            },
          }
        }
      />
    </Modal>
    <div
      className="mdl-layout__content full-height"
    >
      <div
        className="mdl-grid mdl-grid--no-spacing full-height"
      >
        <div
          className="mdl-cell mdl-cell--2-col-desktop mdl-cell--hide-tablet mdl-cell--hide-phone request-page__left-side-bar "
        >
          <div
            className="sidebar "
          >
            <Connect(LeftSideBar)
              selectedLink="request page"
            />
          </div>
        </div>
        <div
          className="mdl-cell mdl-cell--9-col-desktop request-page__table-view mdl-cell--8-col-tablet mdl-cell--4-col-phone"
        >
          <div
            className="rp-requests "
          >
            <div
              className="rp-requests__header"
            >
              <RequestPanelHeader
                activeStatus="all"
                fetchRequests={[Function]}
                getRequestsWithLimit={[Function]}
                modalType={null}
                openModal={[Function]}
                openRequestsCount={1}
                pastRequestsCount={1}
                requestsLength={3}
                shouldOpen={false}
                url="?page=1"
              />
            </div>
            <div
              className="rp-table"
            >
              <WithLoading
                closeModal={[Function]}
                fetchRequestsError={null}
                isLoading={false}
                modalType={null}
                openModal={[Function]}
                requests={
                  Array [
                    Object {
                      "arrivalDate": "2018-12-11",
                      "department": "TDD",
                      "departureDate": "2018-12-09",
                      "destination": "Nairobi",
                      "gender": "Female",
                      "id": "xDh20btGz",
                      "manager": "Samuel Kubai",
                      "name": "Amarachukwu Agbo",
                      "origin": "Lagos",
                      "role": "Software Developer",
                      "status": "Open",
                      "userId": "pommyLHJmKrx76A8Slm",
                    },
                    Object {
                      "arrivalDate": "2018-12-11",
                      "department": "TDD",
                      "departureDate": "2018-12-09",
                      "destination": "Nairobi",
                      "gender": "Female",
                      "id": "xDh20btGy",
                      "manager": "Samuel Kubai",
                      "name": "Amarachukwu Agbo",
                      "origin": "Lagos",
                      "role": "Software Developer",
                      "status": "Rejected",
                      "userId": "pommyLHJmKrx76A8Slm",
                    },
                    Object {
                      "arrivalDate": "2018-12-11",
                      "department": "TDD",
                      "departureDate": "2018-12-09",
                      "destination": "Nairobi",
                      "gender": "Female",
                      "id": "xDh20btGx",
                      "manager": "Samuel Kubai",
                      "name": "Amarachukwu Agbo",
                      "origin": "Lagos",
                      "role": "Software Developer",
                      "status": "Approved",
                      "userId": "pommyLHJmKrx76A8Slm",
                    },
                  ]
                }
                shouldOpen={false}
              />
            </div>
            <div
              className="rp-pagination"
            >
              <Pagination
                currentPage={1}
                onPageChange={[Function]}
                pageCount={4}
              />
            </div>
          </div>
        </div>
        <div
          className="mdl-cell mdl-cell--3-col-desktop hide mdl-cell--hide-desktop request-page__right-side-bar mdl-cell--3-col-tablet mdl-cell--4-col-phone"
        >
          <div
            className="notification hide"
          >
            <NotificationPane
              onCloseNotificationPane={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
